### PR TITLE
Azure Blob Storage: we added an option to check an object after write, too

### PR DESCRIPTION
**Pull request by SmitaRKulkarni**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added option to check object exists after writing to Azure Blob Storage, this is controlled by setting `check_objects_after_upload`

### CODE REVIEW

**Comment by jkartseva** [comment]
- Please add a test for the new flag (possibly, by extending `test_storage_azure_blob_storage`).
- Extend the Changelog entry with information that the change is related to Azure Blob storage.

**Review by jkartseva — APPROVED** [review]
Please address the feedback.

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.h [hunk 1]
```diff
@@ -90,6 +90,7 @@ class WriteBufferFromAzureBlobStorage : public WriteBufferFromFileBase
     size_t hidden_size = 0;
 
     std::unique_ptr<TaskTracker> task_tracker;
+    bool check_objects_after_upload;
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.h]: addressed it

File: src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.cpp [hunk 2]
```diff
@@ -263,6 +263,7 @@ std::unique_ptr<AzureObjectStorageSettings> getAzureBlobStorageSettings(const Po
     settings->strict_upload_part_size = config.getUInt64(config_prefix + ".strict_upload_part_size", context->getSettings().azure_strict_upload_part_size);
     settings->upload_part_size_multiply_factor = config.getUInt64(config_prefix + ".upload_part_size_multiply_factor", context->getSettings().azure_upload_part_size_multiply_factor);
     settings->upload_part_size_multiply_parts_count_threshold = config.getUInt64(config_prefix + ".upload_part_size_multiply_parts_count_threshold", context->getSettings().azure_upload_part_size_multiply_parts_count_threshold);
+    settings->check_objects_after_upload = config.getUInt64(config_prefix + ".check_objects_after_upload", false);
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/ObjectStorages/AzureBlobStorage/AzureBlobStorageAuth.cpp]: right, fixed it

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 3]
```diff
@@ -167,6 +169,33 @@ void WriteBufferFromAzureBlobStorage::finalizeImpl()
...
+        bool found = false;
+
+        for (const auto & blob : blobs_list)
+        {
+            if (blob_path == blob.Name)
+            {
+                found = true;
+                break;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp]: updated the check based on this PR (https://github.com/ClickHouse/ClickHouse/pull/65471)

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp [hunk 4]
```diff
@@ -167,6 +169,25 @@ void WriteBufferFromAzureBlobStorage::finalizeImpl()
...
+            auto blob_client = blob_container_client->GetBlobClient(blob_path);
+            blob_client.GetProperties();
+            return;
+        }
+        catch (const Azure::Storage::StorageException & e)
+        {
+            if (e.StatusCode == Azure::Core::Http::HttpStatusCode::NotFound)
+                throw Exception(
+                        ErrorCodes::AZURE_BLOB_STORAGE_ERROR,
+                        "Object {} not uploaded to azure blob storage, it's a bug in Azure Blob Storage or its API.",
```
Comment (jkartseva) [inline src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:186]: > ... a bug in Azure Blob Storage or its API.

What is the history behind this PR? Did we encounter a bug at some point?

**Review by jkartseva — APPROVED** [review]
LGTM. As an idea, errors in external APIs can be emulated by injecting [FailPoints](https://github.com/ClickHouse/ClickHouse/blob/master/src/Common/FailPoint.h)

File: src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp (hunk 4 above)
Comment (SmitaRKulkarni) [reply src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp:186]: discussed offline

File: src/Core/SettingsChangesHistory.cpp [hunk 5]
```diff
@@ -89,6 +89,7 @@ static std::initializer_list<std::pair<ClickHouseVersion, SettingsChangesHistory
               {"input_format_csv_deserialize_separate_columns_into_tuple", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_csv_try_infer_strings_from_quoted_tuples", true, true, "A new way of how interpret tuples in CSV format was added."},
               {"input_format_json_ignore_key_case", false, false, "Ignore json key case while read json field from string."},
+              {"azure_check_objects_after_upload", false, false, "Check each uploaded object to azure blob storage with head request to be sure that upload was successful"},
```
Comment (jkartseva) [inline src/Core/SettingsChangesHistory.cpp:92]: It should be a `24.7` setting as of now.

**Review by jkartseva — APPROVED** [review]
LGTM. Sorry, I did not merge it because some tests were failing (due to other reasons), and this PR got stuck.

**Comment by SmitaRKulkarni** [comment]
Unrelated failure : 
Stateless tests (tsan, s3 storage) [2/3] - 00653_verification_monotonic_data_load - https://github.com/ClickHouse/ClickHouse/issues/68626

**Pull request merged**